community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/storage/paged/PagedIndexInput.java

```diff
@@ -0,0 +1,441 @@
+            bytesRead += toRead;
+        }
+    }
+
+    @Override
+    public void seek( long pos ) throws IOException
+    {
+        long newPageId = pageId( pos );
+        int newOffset = pageOffset( pos );
+
+        if ( isEOF( newPageId, newOffset ) )
+        {
+            throw new EOFException( "seek past EOF: " + this );
+        }
+
+        currentPageId = newPageId;
+        currentPageOffset = newOffset;
+    }
+
+    @Override
+    public void skipBytes( long numBytes ) throws IOException
+    {
+        long overflowingOffset = currentPageOffset + numBytes;
+        long newPageId = currentPageId + overflowingOffset / pageSize;
+        int newOffset = (int) (overflowingOffset % pageSize);
+
+        if ( isEOF( newPageId, newOffset ) )
+        {
+            throw new EOFException( "seek past EOF: " + this );
+        }
+
+        currentPageId = newPageId;
+        currentPageOffset = newOffset;
+    }
+
+    /**
+     * @return the logical position of this slice, 0 being start of slice
+     */
+    @Override
+    public long getFilePointer()
+    {
+        return currentPageId * pageSize + currentPageOffset - startPosition;
+    }
+
+    @Override
+    public final long length()
+    {
+        return length;
+    }
+
+    @Override
+    public final PagedIndexInput clone()
+    {
+        PagedIndexInput slice = slice( toString(), 0, length() );
+        slice.currentPageId = currentPageId;
+        slice.currentPageOffset = currentPageOffset;
+        return slice;
+    }
+
+    /**
+     * Creates a slice of this index input, with the given description, offset,
+     * and length. The slice is seeked to the beginning.
+     */
+    @Override
+    public final PagedIndexInput slice( String desc, long offset, long length )
+    {
+        // Implementation notes:
+        // This is used for both slice() and clone()
+        // Lucene does not close these child inputs; it only closes the
+        // root one. See PagedIndexInputCloningTest for details.
+        if ( offset < 0 || length < 0 || offset + length > this.length )
+        {
+            throw new IllegalArgumentException(
+                    "slice() " + desc + " out of bounds: offset=" + offset + ",length=" + length + ",fileLength=" +
+                            this.length + ": " + this );
+        }
+
+        try
+        {
+            return new PagedIndexInput( resources.cloneResources(), desc, startPosition + offset, pageSize, length );
+        }
+        catch ( IOException e )
+        {
+            // Can't throw checked IOException due to the Lucene API
+            throw new UncheckedIOException( e );
+        }
+    }
+
+    @Override
+    public final void close() throws IOException
+    {
+        resources.close( this );
+    }
+
+    private void boundsCheck() throws EOFException
+    {
+        if ( cursor.checkAndClearBoundsFlag() )
+        {
+            throw new EOFException( "read past EOF: " + this );
```

Comment:
this exception message is a bit misleading, would be good to include positions, file name etc that   caused this as well

Reply:
+1